warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/job/ShardedTableMapFile.java

```diff
@@ -68,6 +93,102 @@ public static void setupFile(Configuration conf) throws IOException, URISyntaxEx
         return locations;
     }
     
+    public static void validateShardIdLocations(Configuration conf, String tableName, int daysToVerify, Map<Text,String> shardIdToLocation) {
+        ShardIdFactory shardIdFactory = new ShardIdFactory(conf);
+        // assume true unless proven otherwise
+        boolean isValid = true;
+        for (int daysAgo = 0; daysAgo <= daysToVerify; daysAgo++) {
+            long inMillis = System.currentTimeMillis() - (daysAgo * DateUtils.MILLIS_PER_DAY);
+            String datePrefix = DateHelper.format(inMillis);
+            int expectedNumberOfShards = shardIdFactory.getNumShards(datePrefix);
+            boolean shardsExist = shardsExistForDate(shardIdToLocation, datePrefix, expectedNumberOfShards);
+            if (!shardsExist) {
+                log.warn("Shards for " + datePrefix + " for table " + tableName + " do not exist!");
+                isValid = false;
+                continue;
+            }
+            boolean shardsAreBalanced = shardsAreBalanced(shardIdToLocation, datePrefix);
+            if (!shardsAreBalanced) {
+                log.warn("Shards for " + datePrefix + " for table " + tableName + " are not balanced!");
+                isValid = false;
+            }
+        }
+        if (!isValid) {
+            throw new IllegalStateException("All of today's shards have not been created or are not balanced for " + tableName +"."
+                            + "Run bin/ingest/create_shards_since.sh yyyymmdd, substitute yyyymmdd with today's date. "
+                            + "Ensure that the normal mechanism for creating shards is running - e.g, crontab");
```

Comment:
Couldn't the IllegalStateException message here be misleading, since technically the exception might be thrown based on the non-existent and/or unbalanced shards from some day other than today's? That is, depending on 'daysToVerify'.

And maintenance-wise, instead of referencing a specific script name/parameters/crontab/etc (all of which are environment-specific and subject to change), it may be better just to say something like "Shard validation failed for *tableName*. Please ensure that shards have been generated. Check log for details about the dates in question". Or something similar. And maybe also change the log messages above to ERROR rather than WARN, since we're saying that this represents an invalid system state

All in all though, everything else LGTM. The validation is a cool feature. I built, deployed, and tested your branch locally with no issues.

Reply:
That makes sense, updated with recommended changes

Reply:
@ivakegg, @brianloss, this looks good to me. Do you guys see any reason not to go ahead and merge this?